src/shared/exec-util.c

```diff
@@ -34,34 +34,49 @@
 /* Put this test here for a lack of better place */
 assert_cc(EAGAIN == EWOULDBLOCK);
 
-int locate_executable(const char *path, char **ret_executable) {
+int open_executable(const char *path, char **ret_executable) {
+        int r;
+
         if (path_is_absolute(path)) {
+                _cleanup_close_ int fd = -1;
                 char *t;
 
+                fd = open(path, O_PATH|O_CLOEXEC);
```

Comment:
why O_PATH? regular should be fine too, no?

Reply:
Perhaps you might want to execute a binary for which you do not have read access?

Reply:
Well, why not? O_PATH should be a itsy bitsy tiny bit faster.

Reply:
hmm, i wonder what the semantics of fexecve() actually are: is the specified fd itself used for mmap()ing things later? if so, O_PATH shouldn't work at all. Or is the specified file open()ed again, akin to opening /proc/self/fd/$fd? glibc says that on old kernels a fallback is in place for invoking the binary via /proc/self/fd/$fd anyway, so we know that on old kernels things are reopened, and thus O_PATH should be OK. But on new kernels?

Reply:
I added a comment to the function based on @floppym's comment now.

Reply:
Our comments crossposted... I think it must open it again, since I'm able to fexecve a binary I have no read access to.

Reply:
With newer kernels, fexecve() uses the execveat() system call with an empty path and the AT_EMPTY_PATH flag set. The execveat(2) man page says this:
```
       AT_EMPTY_PATH
              If  pathname is an empty string, operate on the file referred to
              by dirfd (which may have been obtained using the open(2)  O_PATH
              flag).
```